x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/ShapeValuesSourceType.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.search.aggregations.support;
+
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.script.AggregationScript;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.FieldContext;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.xcontent.ToXContentFragment;
+
+import java.io.IOException;
+
+public abstract class ShapeValuesSourceType<T extends ToXContentFragment> implements Writeable, ValuesSourceType {
+
```

Comment:
Is this generic really needed? T seems unused here.

Reply:
Looks like a failed attempt to make generic code for the two source types. But now with the ElasticPoint interface, most of my generics are possible to delete, so I've been doing a big cleanup. It's getting simpler and simpler, which feels like the right direction for sure.

Reply:
Looking into the ElasticPoint interface, maybe a better name would we Point2d?

Reply:
Indeed. I'm happy with all suggestions, as `ElasticPoint` was a really wild guess. Interestingly, as I've used it for a few days now, I've started to get used to the name, and started to think of it is as precursor to something involving CRS. So not just 2D, but 2 or 3D, and geo and cartesian, and in fact then any CRS. For this we would simply add a crs attribute to the interface, and either add an option z coordinate, or change the getX and getY into `double[] getCoordinates()`, or even move in a JTS direction with a `Coordinate` class. Then the Point class is simply tying together a Coordinate class to a CRS class.

For me, right now, simply having a single interface (named anything sensible) makes a _HUGE_ difference to the complexity of bring cartesian up to geo.

Reply:
Both concerns were addressed in the previous PR which is now merged. https://github.com/elastic/elasticsearch/pull/89388